Suica自動販売機 STEP5まで実装

ima1zumi opened the pull request:
STEP5まで実装しました。ご確認お願いします🙇‍♂️

ima1zumi committed:
step1: ジュースの管理を実装

ima1zumi committed:
step2: 購入機能実装

ima1zumi committed:
step3: レッドブルと水を購入可能にする

ima1zumi committed:
step4: Suicaに年齢・性別を保存できるようにした

ima1zumi committed:
buyをpurchaseに変更

ima1zumi committed:
不要なSuica.newを削除

ima1zumi committed:
step5: 購入履歴を実装

ima1zumi committed:
配列のレイアウトを統一

ima1zumi committed:
user_sexがISO 5218準拠のコメント追加

ima1zumi committed:
step0: Suicaのチャージ機能を実装

JunichiIto commented:
@ima1zumi 修正が完了したらこのPRにコメントください〜。（ブートキャンプのイベントページにコメントを入れるとキャッチボールするたびにコメントが伸びていってしまうので）

ima1zumi commented:
@JunichiIto 
動画レビューありがとうございます！後ほど確認します。

ima1zumi committed:
user_sex をシンボルの :male, :female に修正

ISO 5218に合わせて数字で登録していたが要件にないためシンボルでもつように修正

ima1zumi committed:
purchase_history のメソッド名, 引数を修正

何をするのか分かりにくいメソッド名だったためsave_purchase_historyにメソッド名を修正
引数にインスタンス変数を渡していたがオブジェクトで渡して問題ないためオブジェクトで渡すよう修正

ima1zumi committed:
mapをkeysに修正

ima1zumi committed:
step5のテストを修正

time周りのテストが正しくテストできていなかったので修正
test_step_5_ジュース名を渡すと販売履歴を取得できる でレッドブルを購入しているが、購入履歴を確認するテストがなかったので追加

ima1zumi committed:
レイアウト, 不要な空白, 改行の修正

クラスメソッドをinitializeより前に移動, 不要な空白と改行を削除

ima1zumi commented:
修正完了しました！動画レビュー分かりやすかったです、ありがとうございます。

`test_step_5_ジュース名を渡すと販売履歴を取得できる` の行数が長くなってしまったことに少し悩みました。
メソッドは短い方がいいという認識なのですが、テストのメソッドも短いほうがよいのでしょうか？

ご確認よろしくお願いします。

JunichiIto commented:
@ima1zumi 修正ありがとうございます！疑問点も含めて後日レビューしてみるのでしばらくお待ちください〜。

JunichiIto commented on lib/suica.rb:
```diff
@@ -1,2 +1,18 @@
+  attr_accessor :balance, :user_age, :user_sex
+
+  def initialize(money, user_age, user_sex)
+    @balance = money >= 100 ? money : nil
```
ここは設計として悩ましいですね。
99円でnewして、そのSuicaで買おうとすると、`balance`が`nil`になってるせいでたぶん例外が出ると思います。
例外が出るくらいなら、なるべく早い段階で「止まれ止まれ〜！！！」と例外を投げてあげた方が良いです。

なので、たとえばこんな感じでしょうか？

```ruby
raise ArguemntError, "moneyは100円以上の金額を指定してください" unless money >= 100
@balance = money
```

いずれにせよmoneyが100円以上か100円未満かでnewの振るまいが変わるので、それを検証するテストコードも必要になりますね。

ただ、ぶっちゃけ、僕の中では`new`したタイミングでは金額の指定ができず、常に0円でスタートにしようと思ってました。
yokkiさんが金額指定可能な形で実装してきたので、こんなふうになっていますが。

なので、そもそも問題文に書いてない仕様は「どうしたらいいですか？」と出題者に確認するのがベストだと思います。
仕事でも仕様がハッキリしない部分は、プログラマが想像で作るよりもちゃんと事前に確認した方がよいですし。

ima1zumi replied:
反応が遅くなってしまってすみません🙇‍♂️ご確認ありがとうございます！！
修正しましたので、ご確認お願いします。

---

> 99円でnewして、そのSuicaで買おうとすると、balanceがnilになってるせいでたぶん例外が出ると思います。
例外が出るくらいなら、なるべく早い段階で「止まれ止まれ〜！！！」と例外を投げてあげた方が良いです。

なるほど。。そこで例外になってしまうのは良くないですね🧐
100円未満の金額をチャージした時に`ArgumentError`を発生させるようにしました。

> なので、そもそも問題文に書いてない仕様は「どうしたらいいですか？」と出題者に確認するのがベストだと思います。
仕事でも仕様がハッキリしない部分は、プログラマが想像で作るよりもちゃんと事前に確認した方がよいですし。

仕様がよくわからない時は仕様を決めている人に確認するようにします！

JunichiIto reviewed (commented):
1点だけコメントしました！それ以外は問題ありません 🙆‍♂️ 
直し方は「こう」と1つに決まるものではないのですが、 @ima1zumi さんの中で「これだ！」と思う設計に修正してみてください〜。

ima1zumi committed:
100円未満のチャージを行った場合ArgumentErrorを返すよう修正

修正前は100円未満のチャージを行った場合nilを返していた。
nilのまま購入処理を行うと例外が発生するためチャージする段階でArgumentErrorを発生させるよう修正した。
合わせてinitializeとchargeで処理内容が重複していたためchargeでのみチャージできるようにした。

JunichiIto reviewed (approved):
@ima1zumi 修正お疲れ様でした。
そうですね、修正後のコードも1つの実装案になると思います！

ただ、newのときに金額が指定できないと、newしてからchargeというふうに、テストコードの手間が1ステップ増えるんですよね。
その点はyokkiさん方式は悪くなったかなーとも思いました。

あと、`Suica.new(18, :female)`のように書くと「18って何の数字？」と思われそうなので、たとえば`Suica.new(age: 18, sex: :female)`のように書いた方がわかりやすいかも、と思いました。（`Suica.new(200, 18, :female)`のときは特に問題を感じなかったのですが）

ただ、このあたりは「何が正解か」という答えがカッチリ決まるわけではないので、「そういう観点もありますよ」ということだけを伝えて終わりにしておきます。

他に何か疑問点等がなければこれで合格（Approve）、とします〜。
どうもお疲れさまでした！

実際にマージすることはできないので、このPRはクローズさせてもらいますね 🙏
